modules/tensor_mechanics/src/postprocessors/InteractionIntegral.C

```diff
@@ -0,0 +1,346 @@
+  {
+    q_avg_seg =
+        (_crack_front_definition->getCrackFrontForwardSegmentLength(_crack_front_point_index) +
+         _crack_front_definition->getCrackFrontBackwardSegmentLength(_crack_front_point_index)) /
+        2.0;
+  }
+
+  Real eq = term1 + term2 - term3 + term4;
+
+  if (_has_symmetry_plane)
+    eq *= 2.0;
+
+  return eq / q_avg_seg;
+}
+
+Real
+InteractionIntegral::computeIntegral()
+{
+  Real sum = 0;
+
+  // calculate phi and dphi for this element
+  FEType fe_type(Utility::string_to_enum<Order>("first"),
+                 Utility::string_to_enum<FEFamily>("lagrange"));
+  const unsigned int dim = _current_elem->dim();
+  UniquePtr<FEBase> fe(FEBase::build(dim, fe_type));
+  fe->attach_quadrature_rule(_qrule);
+  _phi_curr_elem = &fe->get_phi();
+  _dphi_curr_elem = &fe->get_dphi();
+  fe->reinit(_current_elem);
+
+  // calculate q for all nodes in this element
+  _q_curr_elem.clear();
+  unsigned int ring_base = (_q_function_type == "TOPOLOGY") ? 0 : 1;
+
+  for (unsigned int i = 0; i < _current_elem->n_nodes(); ++i)
+  {
+    Node * this_node = _current_elem->get_node(i);
+    Real q_this_node;
+
+    if (_q_function_type == "GEOMETRY")
+      q_this_node = _crack_front_definition->DomainIntegralQFunction(
+          _crack_front_point_index, _ring_index - ring_base, this_node);
+    else if (_q_function_type == "TOPOLOGY")
+      q_this_node = _crack_front_definition->DomainIntegralTopologicalQFunction(
+          _crack_front_point_index, _ring_index - ring_base, this_node);
+
+    _q_curr_elem.push_back(q_this_node);
+  }
+
+  for (_qp = 0; _qp < _qrule->n_points(); _qp++)
+    sum += _JxW[_qp] * _coord[_qp] * computeQpIntegral();
+  return sum;
+}
+
+void
+InteractionIntegral::computeAuxFields(const MooseEnum sif_mode,
+                                      RankTwoTensor & aux_stress,
+                                      RankTwoTensor & grad_disp)
+{
+  RealVectorValue k(0);
+  if (sif_mode == "KI")
+    k(0) = 1;
+
+  else if (sif_mode == "KII")
+    k(1) = 1;
+
+  else if (sif_mode == "KIII")
+    k(2) = 1;
+
+  Real t = _theta;
+  Real t2 = _theta / 2;
+  Real tt2 = 3 * _theta / 2;
+  Real st = std::sin(t);
+  Real ct = std::cos(t);
+  Real st2 = std::sin(t2);
+  Real ct2 = std::cos(t2);
+  Real stt2 = std::sin(tt2);
+  Real ctt2 = std::cos(tt2);
+  Real ct2sq = std::pow(ct2, 2);
+  Real ct2cu = std::pow(ct2, 3);
+  Real sqrt2PiR = std::sqrt(2 * libMesh::pi * _r);
+
+  // Calculate auxiliary stress tensor
+  Real s11 = 1 / sqrt2PiR * (k(0) * ct2 * (1 - st2 * stt2) - k(1) * st2 * (2 + ct2 * ctt2));
+  Real s22 = 1 / sqrt2PiR * (k(0) * ct2 * (1 + st2 * stt2) + k(1) * st2 * ct2 * ctt2);
+  Real s12 = 1 / sqrt2PiR * (k(0) * ct2 * st2 * ctt2 + k(1) * ct2 * (1 - st2 * stt2));
+  Real s13 = -1 / sqrt2PiR * k(2) * st2;
+  Real s23 = 1 / sqrt2PiR * k(2) * ct2;
+  // plain stress
+  // Real s33 = 0;
+  // plain strain
+  Real s33 = _poissons_ratio * (s11 + s22);
+
+  aux_stress(0, 0) = s11;
```

Comment:
I'd prefer to get rid of these intermediate variables (s11, etc) and just directly compute and assign to the components of the stress tensor above:
`aux_stress(0, 0) = 1.0 / sqrt2PiR * (k(0) * ct2 * (1.0 - st2 * stt2) - k(1) * st2 * (2.0 + ct2 * ctt2));`
At the end, then do the symmetric assignments:
`aux_stress(2, 0) = aux_stress(0, 2);`